File: mobile/mglDriverApp/src/mgl/fleetpay-qr.ts
export type FleetpayQrApiFields = {
  mid: string;
  terminalId: string;
  amountPaise: number;
  expiryEpoch: number;
  sign: string;
};

export type FleetpayQrPayload = FleetpayQrApiFields & {
  txnId: string;
  merchantName?: string;
  currency?: string;
};

export function parseFleetpayPayUri(raw: string): FleetpayQrPayload | null {
  const s = raw.trim();
  if (!/^fleetpay:\/\//i.test(s)) return null;
  let url: URL;
  try {
    url = new URL(s);
  } catch {
    return null;
  }
  const pathNorm = url.pathname.replace(/\/+$/, '').toLowerCase();
  const payPathOk = pathNorm === '' || pathNorm.endsWith('/pay');
  const payHostOk = url.hostname.toLowerCase() === 'pay';
  if (!payHostOk && !payPathOk) return null;
  const txnId = url.searchParams.get('txn') ?? '';
  const mid = url.searchParams.get('mid') ?? '';
  const terminalId = url.searchParams.get('tid') ?? '';
  const am = url.searchParams.get('am');
  const exp = url.searchParams.get('exp');
  const sign = url.searchParams.get('sign') ?? '';
  const mn = url.searchParams.get('mn');

  if (!txnId || !mid || !terminalId || !sign || am == null || exp == null) return null;

  const amountPaise = Number(am);
  const expiryEpoch = Number(exp);
  if (!Number.isFinite(amountPaise) || !Number.isFinite(expiryEpoch)) return null;

  let merchantName: string | undefined;
  if (mn) {
    try {
      merchantName = decodeURIComponent(mn.replace(/\+/g, ' '));
    } catch {
      merchantName = mn;
    }
  }

  return {
    txnId,
    mid,
    terminalId,
    amountPaise,
    expiryEpoch,
    sign,
    merchantName,
    currency: url.searchParams.get('cu') ?? undefined,
  };
}

export function paiseToInrDisplay(amountPaise: number): string {
  return (amountPaise / 100).toLocaleString('en-IN', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}

File: flutter-sdk/lib/src/fleetpay_qr.dart
import 'fleet_live_models.dart';

/// Same demo URI as iOS [DriverFleetQr.simulatedUri].
const String kSimulatedFleetpayUri =
    'fleetpay://pay?txn=SIMTXN01&mid=DEMOMID&tid=DEMOTID&am=67200&exp=9999999999&sign=demosign&mn=Demo%20CNG%20Station';

String normVrnPublic(String v) =>
    v.trim().replaceAll(RegExp(r'\s+'), ' ').toUpperCase();

/// Parses `fleetpay://pay?...` URIs; mirrors Android [parseFleetpayPayUri].
ParsedFleetpayQr? parseFleetpayPayUri(String raw) {
  final s = raw.trim();
  if (!RegExp(r'^fleetpay://', caseSensitive: false).hasMatch(s)) {
    return null;
  }
  Uri uri;
  try {
    uri = Uri.parse(s);
  } catch (_) {
    return null;
  }
  final pathNorm = uri.path.toLowerCase().replaceAll(RegExp(r'^/+|/+$'), '');
  final payPathOk = pathNorm.isEmpty || pathNorm.endsWith('/pay');
  final host = (uri.host).toLowerCase();
  final payHostOk = host == 'pay';
  if (!payHostOk && !payPathOk) return null;

  String? qp(String name) => uri.queryParameters[name];

  final txnId = qp('txn');
  final mid = qp('mid');
  final terminalId = qp('tid');
  final am = qp('am');
  final exp = qp('exp');
  final sign = qp('sign');
  if (txnId == null ||
      mid == null ||
      terminalId == null ||
      am == null ||
      exp == null ||
      sign == null) {
    return null;
  }
  final amountPaise = double.tryParse(am)?.toInt();
  final expiryEpoch = double.tryParse(exp)?.toInt();
  if (amountPaise == null || expiryEpoch == null) return null;

  return ParsedFleetpayQr(
    txnId: txnId,
    mid: mid,
    terminalId: terminalId,
    amountPaise: amountPaise,
    expiryEpoch: expiryEpoch,
    sign: sign,
    merchantName: qp('mn'),
    currency: qp('cu'),
  );
}

String paiseToInrDisplay(int amountPaise) {
  final inr = amountPaise / 100.0;
  return inr.toStringAsFixed(2);
}

bool validIndianMobile10(String digitsOnly) {
  final d = digitsOnly.replaceAll(RegExp(r'\D'), '');
  if (d.length < 10) return false;
  final last10 = d.length == 10 ? d : d.substring(d.length - 10);
  final first = int.tryParse(last10[0]);
  return first != null && first >= 6 && first <= 9;
}
